File: src/ShowyEdge/objc/MenuController.m
#import "MenuController.h"
#import "PreferencesManager.h"
#import "PreferencesWindowController.h"

@interface MenuController ()

@property(weak) IBOutlet PreferencesWindowController* preferencesWindowController;
@property NSStatusItem* item;

@end

@implementation MenuController

- (void)show {
  if (!self.item) {
    self.item = [NSStatusBar.systemStatusBar statusItemWithLength:NSSquareStatusItemLength];
    self.item.button.image = [NSImage imageNamed:@"menu"];

    NSMenu* menu = [[NSMenu alloc] initWithTitle:@"ShowyEdge"];

    NSString* version = [[NSBundle mainBundle] infoDictionary][@"CFBundleVersion"];
    [menu addItemWithTitle:[NSString stringWithFormat:@"ShowyEdge %@", version]
                    action:nil
             keyEquivalent:@""];

    [menu addItem:[NSMenuItem separatorItem]];

    {
      NSMenuItem* menuItem = [[NSMenuItem alloc] initWithTitle:@"Preferences..."
                                                        action:@selector(showPreferences:)
                                                 keyEquivalent:@""];
      menuItem.target = self;
      [menu addItem:menuItem];
    }

    [menu addItem:[NSMenuItem separatorItem]];

    {
      NSMenuItem* menuItem = [[NSMenuItem alloc] initWithTitle:@"Quit ShowyEdge"
                                                        action:@selector(terminate:)
                                                 keyEquivalent:@""];
      menuItem.target = self;
      [menu addItem:menuItem];
    }

    self.item.menu = menu;
  }

  self.item.visible = PreferencesManager.showIconInMenubar;
}

- (void)showPreferences:(id)sender {
  [self.preferencesWindowController show];
}

- (void)terminate:(id)sender {
  [NSApp terminate:self];
}

@end

File: src/ShowyEdge/swift/MenuController.swift
import AppKit

class MenuController: NSObject {
  static let shared = MenuController()
  var statusItem: NSStatusItem?

  public func show() {
    if statusItem == nil {
      statusItem = NSStatusBar.system.statusItem(
        withLength: NSStatusItem.squareLength
      )

      statusItem?.button?.image = NSImage(named: "menu")

      let menu = NSMenu(title: "ShowyEdge")

      let version = Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as! String
      menu.addItem(withTitle: "ShowyEdge \(version)", action: nil, keyEquivalent: "")

      menu.addItem(NSMenuItem.separator())

      let settingsMenuItem = NSMenuItem(
        title: "Settings...",
        action: #selector(showSettings),
        keyEquivalent: "")
      settingsMenuItem.target = self
      menu.addItem(settingsMenuItem)

      menu.addItem(NSMenuItem.separator())

      menu.addItem(
        withTitle: "Quit ShowyEdge",
        action: #selector(NSApplication.shared.terminate),
        keyEquivalent: ""
      )

      statusItem?.menu = menu
    }

    //
    // Set visibility
    //

    statusItem?.isVisible = UserSettings.shared.showMenu
  }

  @objc private func showSettings(sender _: AnyObject?) {
    SettingsWindowManager.shared.show()
  }
}
